### PR TITLE
fix(cli/self-update): fix Nushell-related suggestions and scripts

### LINKED ISSUE #4242: Nushell environment script (`env.nu`) is using the wrong string
### Verification

- [x] I searched for recent similar issues at https://github.com/rust-lang/rustup/issues?q=is%3Aissue+is%3Aopen%2Cclosed and found no duplicates.
- [x] I am on the latest version of Rustup according to https://github.com/rust-lang/rustup/tags and am still able to reproduce my issue.

### Problem

I just installed rustup via `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh`

I use nushell, so when I next opened nushell it was very broken and both `nushell/env.nu` and `nushell/config.nu` contained `source "$HOME/.cargo/env.nu`

I fixed this with a `source ~/.cargo/env.nu`. You could equally use something like `source ($nu.home-path | join `.cargo/env.nu)`

cargo still couldn't be found, and I noticed the `.cargo/env.nu` was this:
```nushell
if ("$HOME/.cargo/bin" not-in ($env.Path | split row (char esep))) {
  $env.Path = ($env.Path | prepend "$HOME/.cargo/bin")
}
```

Which resulted in a path of `PATH=/home/user/$HOME/.cargo/bin:/home/user/.local/bin: ...`

I haven't investigated what has caused this yet, but I would be happy to do so a bit later

### Steps

1. Prepare a system with nushell but not cargo
2. Install cargo via rustip (`curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh`)
3. Open nushell and observe its errors

### Possible Solution(s)

I assume some string interpolation has changed somewhere?

### Notes

_No response_

### Rustup version

```console
rustup 1.28.1 (f9edccde0 2025-03-05)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.87.0-nightly (b74da9613 2025-03-06)`
```

### Installed toolchains

```console
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/user/.rustup

installed toolchains
--------------------
nightly-x86_64-unknown-linux-gnu (active, default)

active toolchain
----------------
name: nightly-x86_64-unknown-linux-gnu
active because: it's the default toolchain
installed targets:
  x86_64-unknown-linux-gnu
```

### OS version

```console
ubuntu 24.04
```

### PR DESCRIPTION
**Pull request by rami3l**
Supersedes #4230. Closes #4242.

PS: Thanks for your contribution, @yethal! I've made you the co-author of ~~two~~ three main commits in this PR.

@alluring-mushroom @yethal thanks again for your report! I don't use Nushell personally and I'm not sure if this has covered all the corner cases (it would be difficult to test automatically due to the amount of different setups we need to support). Would you mind trying out this patch and see if the fix works for you?

### CODE REVIEW

File: src/cli/self_update/shell.rs [hunk 1]
```diff
@@ -256,7 +260,20 @@ impl UnixShell for Fish {
     }
 }
 
-struct Nu;
+pub(super) struct Nu;
+
+impl Nu {
+    pub(super) fn translate_cargo_home(cargo_home: &mut String) {
```
Comment (rami3l) [inline src/cli/self_update/shell.rs]: It looks like `cargo_home_str` should probably be a method in the `UnixShell` trait with a default implementation.

Comment (rami3l) [reply src/cli/self_update/shell.rs]: According to @alluring-mushroom's report, it could be that the content of `env.nu` is also wrong because it contains `$HOME`: the `UnixShell` trait might need a more thorough rework.

**Comment by Yethal** [comment]
@rami3l Just ran this in a VM
Preinstall prompt:
```
This path will then be added to your PATH environment variable by
modifying the profile files located at:

  /home/vagrant/.profile
  /home/vagrant/.bashrc
  /home/vagrant/.config/nushell/env.nu
  /home/vagrant/.config/nushell/config.nu
```
Post install prompt:
```
...
To configure your current shell, you need to source
the corresponding env file under $HOME/.cargo.

This is usually done by running one of the following (note the leading DOT):
. "$HOME/.cargo/env"            # For sh/bash/zsh/ash/dash/pdksh
source "$HOME/.cargo/env.fish"  # For fish
source $"($nu.home-path)/.cargo/env.nu"  # For nushell
```

Generated env.nu:
```
if ("$HOME/.cargo/bin" not-in ($env.Path | split row (char esep))) {
  $env.Path = ($env.Path | prepend "$HOME/.cargo/bin")
}
```

$env.PATH contents:
```
╭────┬──────────────────────────╮
│  0 │ $HOME/.cargo/bin         │
│  1 │ /home/vagrant/.cargo/bin │
│  2 │ /usr/local/sbin          │
│  3 │ /usr/local/bin           │
│  4 │ /usr/sbin                │
│  5 │ /usr/bin                 │
│  6 │ /sbin                    │
│  7 │ /bin                     │
│  8 │ /usr/games               │
│  9 │ /usr/local/games         │
│ 10 │ /snap/bin                │
╰────┴──────────────────────────╯
```
Unfortunately generated env.nu still uses posix-like `$HOME` syntax instead of nu-specific `$"($nu.home-path)/.cargo/bin"` so this still doesn't work as advertised (but no longer breaks existing configs so there's that)
Other than that you may want to consider only appending to `config.nu` since `env.nu` is about to be deprecated

**Comment by rami3l** [comment]
@yethal Thanks for your quick response. As per https://github.com/rust-lang/rustup/pull/4265#discussion_r2010452958 this is the expected behavior (for now); will fix.

**Comment by alluring-mushroom** [comment]
@Yethal 
> `env.nu` is about to be deprecated 

Thank you for this, I hadn't been following the recent changes.

@rami3l if you look at the [documentation for the next release](https://www.nushell.sh/book/configuration.html#using-constants) 

It shows out nushell intends for it to be done in the future

File: src/cli/self_update/shell.rs (hunk 1 above)
Comment (rami3l) [reply src/cli/self_update/shell.rs]: Fixed.

**Comment by rami3l** [comment]
@Yethal @alluring-mushroom Sorry for the back and forth, but I have tried to apply the suggestions from both of you. Would you mind giving this PR another shot?

**Comment by Yethal** [comment]
```
~> $env.PATH
╭───┬──────────────────────────╮
│ 0 │ /home/vagrant/.cargo/bin │
│ 1 │ /usr/local/sbin          │
│ 2 │ /usr/local/bin           │
│ 3 │ /usr/sbin                │
│ 4 │ /usr/bin                 │
│ 5 │ /sbin                    │
│ 6 │ /bin                     │
│ 7 │ /usr/games               │
│ 8 │ /usr/local/games         │
│ 9 │ /snap/bin                │
╰───┴──────────────────────────╯
```
@rami3l We're good. Path is properly appended so the issue is resolved.

For easier multiplatorm support (once that happens) you may want to switch over to using the `path add` function from the Nushell standard library which will take care of deduplicating Path entries for you as well as `path join` for platform specific path separators instead of hardcoding `/`. Code works fine as is though so treat this as a suggestion and not something that needs to be implemented right now.

```
use std/util "path add"
path add ($nu.home-path | path join .cargo bin)
```

**Comment by rami3l** [comment]
@Yethal Thanks for your feedback! I'm glad to see that this PR is working for you.

Regarding your two suggestions, I think since we are handling path joining from Rust's side, it might be better to postpone the change regarding `path join`.

On the other hand, your `path add` suggestion does look like a good improvement that we can integrate right now. Regarding the final `env.nu` template, I'm thinking about something like the following:

```diff
-if ($"{cargo_bin}" not-in ($env.Path | split row (char esep))) {
-  $env.Path = ($env.Path | prepend $"{cargo_bin}")
-}
+use std/util "path add"
+path add $"{cargo_bin}"
```

... I hope that looks reasonable to you. Besides, do you think this `use std/util "path add"` API is stable enough for the inclusion?

**Comment by Yethal** [comment]
> @Yethal Thanks for your feedback! I'm glad to see that this PR is working for you.
> 
> Regarding your two suggestions, I think since we are handling path joining from Rust's side, it might be better to postpone the change regarding `path join`.
> 
> On the other hand, your `path add` suggestion does look like a good improvement that we can integrate right now. Regarding the final `env.nu` template, I'm thinking about something like the following:
> 
> ```diff
> -if ($"{cargo_bin}" not-in ($env.Path | split row (char esep))) {
> -  $env.Path = ($env.Path | prepend $"{cargo_bin}")
> -}
> +use std/util "path add"
> +path add $"{cargo_bin}"
> ```
> 
> ... I hope that looks reasonable to you. Besides, do you think this `use std/util "path add"` API is stable enough for the inclusion?

Yes, this looks reasonable. `path add` is part of Nushell's standard library so I'd say it's stable enough

File: src/cli/self_update/shell.rs [hunk 2]
```diff
@@ -41,14 +41,19 @@ pub(crate) struct ShellScript {
...
+fn cargo_home_str_with_home(home: &str, process: &Process) -> Result<Cow<'static, str>> {
     let path = process.cargo_home()?;
 
     let default_cargo_home = process
         .home_dir()
         .unwrap_or_else(|| PathBuf::from("."))
         .join(".cargo");
     Ok(if default_cargo_home == path {
-        "$HOME/.cargo".into()
+        Cow::Owned(home.to_owned() + "/.cargo")
```
Comment (djc) [inline src/cli/self_update/shell.rs]: Nit: I think it would be clearer to write this as `Cow::Owned(format!("{home}/.cargo"))`.

File: src/cli/self_update/shell.rs [hunk 3]
```diff
@@ -101,21 +96,27 @@ pub(crate) trait UnixShell {
         }
     }
 
+    fn cargo_home_str(&self, process: &Process) -> Result<Cow<'static, str>> {
+        let home = "$HOME";
```
Comment (djc) [inline src/cli/self_update/shell.rs:103]: Nit: don't see much point in the `home` binding in the context of this commit. Suggest inlining it here.

File: src/cli/self_update.rs [hunk 4]
```diff
@@ -578,13 +581,21 @@ pub(crate) async fn install(
...
     };
     #[cfg(not(windows))]
-    let msg = if no_modify_path {
-        format!(
-            post_install_msg_unix_no_modify_path!(),
-            cargo_home = cargo_home
-        )
-    } else {
-        format!(post_install_msg_unix!(), cargo_home = cargo_home)
+    let msg = {
```
Comment (djc) [inline src/cli/self_update.rs]: Nit: suggest skipping the outer scope here.

File: src/cli/self_update/unix.rs [hunk 5]
```diff
@@ -174,15 +174,15 @@ fn remove_legacy_paths(process: &Process) -> Result<()> {
     remove_legacy_source_command(
         format!(
             "export PATH=\"{}/bin:$PATH\"\n",
-            shell::cargo_home_str(process)?
+            Posix.cargo_home_str(process)?
```
Comment (djc) [inline src/cli/self_update/unix.rs:177]: Should we pass some kind of shell into here? Why is it okay to hardcode the shell here?

**Review by djc — COMMENTED** [review]
Good stuff!

File: src/cli/self_update/shell.rs (hunk 3 above)
Comment (rami3l) [reply src/cli/self_update/shell.rs:103]: The point will show up later... It's one of the side effects of rewriting history I think :)

File: src/cli/self_update/unix.rs (hunk 5 above)
Comment (rami3l) [reply src/cli/self_update/unix.rs:177]: In legacy error notes we *assume* POSIX-style paths. You can click to expand the context to get a better view.

File: src/cli/self_update/shell.rs (hunk 2 above)
Comment (rami3l) [reply src/cli/self_update/shell.rs]: Done.

File: src/cli/self_update/shell.rs (hunk 3 above)
Comment (rami3l) [reply src/cli/self_update/shell.rs:103]: The point will show up later in [`908285b` (#4265)](https://github.com/rust-lang/rustup/pull/4265/commits/908285bd067badcd1517999c5cfc0ad6914e0127)... Either way it won't make that much of a difference, so I've changed that according to your suggestion.

File: src/cli/self_update/unix.rs (hunk 5 above)
Comment (djc) [reply src/cli/self_update/unix.rs:177]: Fair enough!

**Review by djc — APPROVED** [review]
